### PR TITLE
Add the option to skip the creation of the bootstrap service in the Load

### PR DESCRIPTION
**Pull request by yyang48**
Balancer Kafka Listener.

This is the fix for the issue: https://github.com/strimzi/strimzi-kafka-operator/issues/4433

Signed-off-by: Yaodong Yang <yyang48@apple.com>

### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

_Please describe your pull request_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md
- [x] Supply screenshots for visual changes, such as Grafana dashboards

Screen Shot 2022-01-20 at 8 53 19 AM

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 1]
```diff
@@ -2053,6 +2054,13 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
             for (GenericKafkaListener listener : loadBalancerListeners) {
                 String bootstrapServiceName = ListenersUtils.backwardsCompatibleBootstrapServiceName(name, listener);
 
+                // When the listener is load balancer and useCreateBootStrapService is false,
+                // we will skip the creation of the external load balancer to save the cost.
+                if (KafkaListenerType.LOADBALANCER == listener.getType() && listener.getConfiguration() != null
+                        && !listener.getConfiguration().getCreateBootstrapService()) {
+                    continue;
+                }
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:2062]: I'm afraid this would be too simple :-/. It skips the creation of the ListenerStatus which we definitely do not want to skip. We would want it at least without the address, but ideally with the address constructed from the broker services / loadbalancers.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 2]
```diff
@@ -790,6 +790,10 @@ public Service generateService() {
         List<Service> services = new ArrayList<>(externalListeners.size());
 
         for (GenericKafkaListener listener : externalListeners)   {
+            if (KafkaListenerType.LOADBALANCER == listener.getType() && listener.getConfiguration() != null && !listener.getConfiguration().getCreateBootstrapService()) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:793]: MAybe this `if` can have some method in https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersUtils.java ... you already seem to be using it in 2 places.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 3]
```diff
@@ -229,6 +230,17 @@ public void setIpFamilies(List<IpFamily> ipFamilies) {
         this.ipFamilies = ipFamilies;
     }
 
+    @Description("Whether to create the bootstrap service or not." +
+            "This field can be used with the `loadBalancer` type listener.")
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:234]: Since we support this only with load balancers, you should also add some validation to https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersValidator.java

**Review by scholzj — COMMENTED** [review]
Thanks for the PR. I left some initial comments.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaTest.java [hunk 4]
```diff
@@ -80,15 +82,27 @@ public void testCertificationAuthorityBuilderAndInts() throws URISyntaxException
...
+        GenericKafkaListenerConfiguration genericKafkaListenerConfiguration = new GenericKafkaListenerConfiguration();
+        genericKafkaListenerConfiguration.setCreateBootstrapService(false);
+
+        listeners.add(new GenericKafkaListenerBuilder()
+                .withName("snd")
+                .withPort(9094)
+                .withType(KafkaListenerType.LOADBALANCER)
+                .withTls(true)
+                .withConfiguration(genericKafkaListenerConfiguration)
+                .build());
```
Comment (scholzj) [inline api/src/test/java/io/strimzi/api/kafka/model/KafkaTest.java:104]: I guess it is fine to keep this ... but TBH it is not really needed. I do not think this option affects the serialization in any significant way which is  what is really tested here.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 1 above)
Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:2062]: Sorry, I don't understand what you mentioned above.  You mean we want to keep the headless service for the bootstrap service, but don't assign a load balancer IP for it? Here is my test example:

If we skip the bootstrap service, I got the following services. The external bootstrap service is skipped.

image

Do we still want to create the external-bootstrap service, but without an External Load Balancer assigned? I guess my confusion is where is the ListenerStatus defined.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java (hunk 2 above)
Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:793]: done.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaTest.java (hunk 4 above)
Comment (yyang48) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaTest.java:104]: Yes, it's just a boolean value, not a complex object. I take a second thought before adding them here next time

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 1 above)
Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: put it in another way, what should the expect output be.

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: When you look at the deployed Kafka cluster, in the status section you should see something like this:

```yaml
    - addresses:
      - host: 192.168.1.72
        port: 9094
      bootstrapServers: 192.168.1.72:9094
      certificates:
      - |
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      type: external
```

That is all generated here on the lines 2077-2084. Your code basiclly just skips this, so there would be no record about this listener in the status. I think we should think about how to handle that. Ideally, we would want there multiple addresses instead of the _per-broker_ load balancers. Not sure how easy that would be. But if nothing else, we would probably want to at least have the information abott the listener without the address.

Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: I got your point now, thanks for the explanation! Let me look at the code again.

Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: So we would like to change the listener status to this (if we skip the creation of the bootstrap service)? use the above example:

``` 
     - addresses:
      bootstrapServers: 192.168.1.72:9094 
      certificates:
      - |
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      type: external
```

It seems like a better idea as it aligns with Apache Kafka's bootstrap services, so it's widely recognized by the kafka users.

Am I correct?

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Yes. I think I would be flexible if it would be too complicated. But ideally we would probably want something like this when the bootstrap service is disabled:

```yaml
    - addresses:
        - host: 192.168.1.72
          port: 9094
        - host: 192.168.1.73
          port: 9094
        - host: 192.168.1.74
          port: 9094
      bootstrapServers: 192.168.1.72:9094,192.168.1.73:9094,192.168.1.74:9094 
      certificates:
        - |
          -----BEGIN CERTIFICATE-----
          ...
          -----END CERTIFICATE-----
      type: external
```

Where `192.168.1.72:9094,192.168.1.73:9094,192.168.1.74:9094 ` are the addresses of the loadbalancers corresponding to the individual brokers.

Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Got it, thanks! It shouldn't be hard. let me try again and update the PR later.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 5]
```diff
@@ -2053,78 +2053,88 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
...
-                            String bootstrapAddress;
-
-                            if (svc.getStatus().getLoadBalancer().getIngress().get(0).getHostname() != null)    {
-                                bootstrapAddress = svc.getStatus().getLoadBalancer().getIngress().get(0).getHostname();
-                            } else {
-                                bootstrapAddress = svc.getStatus().getLoadBalancer().getIngress().get(0).getIp();
-                            }
-
-                            LOGGER.debugCr(reconciliation, "Found address {} for Service {}", bootstrapAddress, bootstrapServiceName);
+                List<String> listenerAddressList = new ArrayList<>(kafkaCluster.getReplicas() + 1);
```
Comment (yyang48) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:2056]: @scholzj I guess I probably need your help to take a look at this section...

I tested a couple of times about my PR and here is what I found:
1. In the new PR, the operator didn't wait for the completion of generating these Load Balancer Addresses, and it continue to generate the broker certificate and even create the pods with an empty Load Balancer address. Therefore the Kafka Broker can't start because of the lack of the Load Balancer address.
2. My suspicious is my understanding of the vert.x future is not correct, do I need a join after get the bootstrap address in line 2080?

Thanks!

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:2056]: Yeah. The way you have it written, the `perListenerFut` will complete either on line 2057 or 2077. The whole block between lines 2082-2137 => it will not wait for it because it will be already complete. If you want, I can try to rewrite it tomorrow and share the code with you to compare it. Or we can have some call to go through it via screen sharing tomorrow (or when you have time). Whatever you prefer.

Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Thank you for your time and the help. I tested locally, it works! Updating the PR now and adding some tests on it.

Comment (yyang48) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: @scholzj do you have any suggestions for the testing? I'm thinking to add the test in this file: https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java. It doesn't have the load balancer listener today, I guess I can add this type and test it with `createBootstrapService` disabled. Or it's better to do it in other places?

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: No, please don't add it there. That parametrized test is quite bad since any test you add there will run in many unrelated combinations. Maybe `KafkaAssemblyOperatorNonParametrizedTest` would be a good place?

**Comment by yyang48** [comment]
@scholzj could you please take another look for this PR? It's ready for the review. Thanks!

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 6]
```diff
@@ -229,6 +230,17 @@ public void setIpFamilies(List<IpFamily> ipFamilies) {
...
+    @JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:235]: Can you set this to NON_DEFAULT? That way it should not always render into the YAML unless it would be set to false (and thus not change the YAMLs which already exists without the value.)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersValidator.java [hunk 7]
```diff
@@ -196,6 +198,19 @@ private static void validateServiceDnsDomain(Set<String> errors, GenericKafkaLis
...
 
+    /**
+     * Validates that createBootstrapService is used only with Load Balancer type listener
+     *
+     * @param errors    List where any found errors will be added
+     * @param listener  Listener which needs to be validated
+     */
+    private static void validateCreateBootstrapService(Set<String> errors, GenericKafkaListener listener) {
+        if (KafkaListenerType.LOADBALANCER != listener.getType()
+                && listener.getConfiguration() != null && !listener.getConfiguration().getCreateBootstrapService())    {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersValidator.java:209]: Only a nit ... but might be easier to read if you split it into two lines:

```java
        if (KafkaListenerType.LOADBALANCER != listener.getType()
                && listener.getConfiguration() != null 
                && !listener.getConfiguration().getCreateBootstrapService())    {
```

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 8]
```diff
@@ -229,6 +230,17 @@ public void setIpFamilies(List<IpFamily> ipFamilies) {
         this.ipFamilies = ipFamilies;
     }
 
+    @Description("Whether to create the bootstrap service or not." +
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:233]: You need to add a space after the `.`. Otherwise in the docs this is then rendered as `Whether to create the bootstrap service or not.This field can be used with the loadBalancer type listener.` without the space between the sentences.

**Review by scholzj — COMMENTED** [review]
I left some nits. But it looks great otherwise. I will also trigger the regression tests.

**Comment by scholzj** [comment]
PS: The regression failure is not related to this PR.

**Comment by scholzj** [comment]
It looks like it needs to update the CRDs:

```
ERROR: Uncommitted changes in derived resources:
M	documentation/modules/appendix_crds.adoc
M	packaging/helm-charts/helm3/strimzi-kafka-operator/crds/040-Crd-kafka.yaml
M	packaging/install/cluster-operator/040-Crd-kafka.yaml
Run the following to add up-to-date resources:
  mvn clean verify -DskipTests -DskipITs \
    && make crd_install \
    && make helm_install \
    && git add packaging/install/ packaging/helm-charts/ documentation/modules/appendix_crds.adoc cluster-operator/src/main/resources/cluster-roles \
    && git commit -s -m 'Update derived resources'
```

**Comment by yyang48** [comment]
> It looks like it needs to update the CRDs:
> 
> ```
> ERROR: Uncommitted changes in derived resources:
> M	documentation/modules/appendix_crds.adoc
> M	packaging/helm-charts/helm3/strimzi-kafka-operator/crds/040-Crd-kafka.yaml
> M	packaging/install/cluster-operator/040-Crd-kafka.yaml
> Run the following to add up-to-date resources:
>   mvn clean verify -DskipTests -DskipITs \
>     && make crd_install \
>     && make helm_install \
>     && git add packaging/install/ packaging/helm-charts/ documentation/modules/appendix_crds.adoc cluster-operator/src/main/resources/cluster-roles \
>     && git commit -s -m 'Update derived resources'
> ```

Yes, I just fixed it. btw, I really like this error message, it even give me the command to fix it :-)

**Review by scholzj — APPROVED** [review]

**Comment by scholzj** [comment]
(ST failures are not related to this PR)

File: api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java [hunk 9]
```diff
@@ -229,6 +230,17 @@ public void setIpFamilies(List<IpFamily> ipFamilies) {
         this.ipFamilies = ipFamilies;
     }
 
+    @Description("Whether to create the bootstrap service or not. " +
+            "This field can be used with the `loadBalancer` type listener.")
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/arraylistener/GenericKafkaListenerConfiguration.java:234]: Maybe it's worth adding that it's created by default (if not specified differently).

**Review by ppatierno — APPROVED** [review]
LGTM. Just a nit.

**Comment by scholzj** [comment]
Thanks for the PR. Great work!
